Add the support of Empty and Non-Empty filter

pacyL2K19 opened the pull request:
# This PR

- Fix #5216 
- Add the empty filter for fields of type: TEXT, EMAIL, FULL_NAME (composite), DATE_TIME, NUMBER, CURRENCY and LINK

Here is the current behavior:
https://www.loom.com/share/1ae2f5ed73c24c3eae6356532f29bfaa?sid=d58da6e4-d18c-446b-833e-00c6429001c2

@FelixMalfait  please have a look, excited to getting your feedback :) 
Not sure this should be a good first issue though 😃

FelixMalfait commented:
Looks great @pacyL2K19! Indeed not a great first issue 😅
@lucasbordeau could you please review it?

martmull commented on packages/twenty-front/src/modules/object-record/record-filter/utils/turnObjectDropdownFilterIntoQueryFilter.ts:
```diff
@@ -24,6 +24,128 @@ export type ObjectDropdownFilter = Omit<Filter, 'definition'> & {
+      break;
+    case 'Relation':
+      emptyRecordFilter = {
+        [correspondingField.name + 'Id']: { is: 'NULL' } as UUIDFilter,
```
create a `RelationFilter` instead using `UUIDFilter`

martmull replied:
I think you should add one filter type dedicated to Relation even though it defines the same type as UUIDFilter. 
I would do: 
```
type RelationFilter = {
  is?: IsFilter;
}
```
This type can be enriched with other filter operations when necessary

pacyL2K19 replied:
Thanks for clarifying, added the `RelationFilter` type

martmull commented on packages/twenty-front/src/modules/object-record/record-filter/utils/turnObjectDropdownFilterIntoQueryFilter.ts:
```diff
@@ -24,6 +24,128 @@ export type ObjectDropdownFilter = Omit<Filter, 'definition'> & {
+  operand: ViewFilterOperand,
+  correspondingField: Pick<Field, 'id' | 'name'>,
+  objectRecordFilters: RecordGqlOperationFilter[],
+  filterType:
+    | 'URL'
+    | 'Address'
+    | 'String'
+    | 'Number'
+    | 'Date'
+    | 'Relation'
+    | 'CompositeString' = 'String',
```
think you should use `Pick<FilterType, 'LINK' | 'ADDRESS' | 'FULL_NAME' | ...>`

lucasbordeau replied:
We should work only with field types here, not introduce a new type

martmull commented on packages/twenty-front/src/modules/object-record/record-filter/utils/turnObjectDropdownFilterIntoQueryFilter.ts:
```diff
@@ -108,6 +253,15 @@ export const turnObjectDropdownFilterIntoQueryFilter = (
+              rawUIFilter.operand,
+              correspondingField,
+              objectRecordFilters,
+              'Number',
```
```suggestion
              rawUIFilter.definition.type,
```

martmull commented on packages/twenty-front/src/modules/object-record/record-filter/utils/turnObjectDropdownFilterIntoQueryFilter.ts:
```diff
@@ -145,6 +299,15 @@ export const turnObjectDropdownFilterIntoQueryFilter = (
+                rawUIFilter.operand,
+                correspondingField,
+                objectRecordFilters,
+                'Relation',
```
```suggestion
                rawUIFilter.definition.type,
```

martmull commented on packages/twenty-front/src/modules/object-record/record-filter/utils/turnObjectDropdownFilterIntoQueryFilter.ts:
```diff
@@ -169,6 +332,14 @@ export const turnObjectDropdownFilterIntoQueryFilter = (
+            applyEmptyFilters(
+              rawUIFilter.operand,
+              correspondingField,
+              objectRecordFilters,
```
```suggestion
              objectRecordFilters,
              rawUIFilter.definition.type,
```

martmull commented on packages/twenty-front/src/modules/object-record/record-filter/utils/turnObjectDropdownFilterIntoQueryFilter.ts:
```diff
@@ -197,6 +368,15 @@ export const turnObjectDropdownFilterIntoQueryFilter = (
+              rawUIFilter.operand,
+              correspondingField,
+              objectRecordFilters,
+              'URL',
```
same

martmull commented on packages/twenty-front/src/modules/object-record/record-filter/utils/turnObjectDropdownFilterIntoQueryFilter.ts:
```diff
@@ -224,6 +404,15 @@ export const turnObjectDropdownFilterIntoQueryFilter = (
+              rawUIFilter.operand,
+              correspondingField,
+              objectRecordFilters,
+              'CompositeString',
```
same

martmull commented on packages/twenty-front/src/modules/object-record/record-filter/utils/turnObjectDropdownFilterIntoQueryFilter.ts:
```diff
@@ -293,6 +482,15 @@ export const turnObjectDropdownFilterIntoQueryFilter = (
+              rawUIFilter.operand,
+              correspondingField,
+              objectRecordFilters,
+              'Address',
```
same

martmull commented on packages/twenty-front/src/modules/object-record/object-filter-dropdown/components/ObjectFilterDropdownOperandSelect.tsx:
```diff
@@ -34,9 +35,25 @@ export const ObjectFilterDropdownOperandSelect = () => {
+      ViewFilterOperand.IsNotEmpty,
+    ].includes(newOperand);
+
     setSelectedOperandInDropdown(newOperand);
```
you can rename if you want, but not mandatory for an approval

martmull reviewed (changes requested):
Hey, thanks for your submission, it is very nice work. I have left some comments about typing

lucasbordeau commented on packages/twenty-front/src/modules/object-record/record-filter/utils/turnObjectDropdownFilterIntoQueryFilter.ts:
```diff
@@ -24,6 +24,128 @@ export type ObjectDropdownFilter = Omit<Filter, 'definition'> & {
+      };
+      break;
+    case 'CompositeString':
+      // eslint-disable-next-line no-case-declarations
```
Use brackets instead `{ ... }`

martmull replied:
prefer not declare fullNameFilters instead of adding the eslint disable comment
```suggestion
      emptyRecordFilter = {
        and: generateILikeFiltersForCompositeFields(
          '',
          correspondingField.name,
          ['firstName', 'lastName'],
          true,
        ),
      };
```

lucasbordeau commented on packages/twenty-front/src/modules/object-record/record-filter/utils/turnObjectDropdownFilterIntoQueryFilter.ts:
```diff
@@ -24,6 +24,128 @@ export type ObjectDropdownFilter = Omit<Filter, 'definition'> & {
+        ['firstName', 'lastName'],
+      );
+      emptyRecordFilter = {
+        not: {
```
Shouldn't we check for NULL here too ?

lucasbordeau commented on packages/twenty-front/src/modules/object-record/record-filter/utils/turnObjectDropdownFilterIntoQueryFilter.ts:
```diff
@@ -24,6 +24,128 @@ export type ObjectDropdownFilter = Omit<Filter, 'definition'> & {
+      break;
+    case 'Address':
+      emptyRecordFilter = {
+        or: [
```
Shouldn't it be an and ? Because it feels like an address field is empty when all fields are empty no ?

martmull commented on packages/twenty-front/src/modules/object-record/record-filter/utils/turnObjectDropdownFilterIntoQueryFilter.ts:
```diff
@@ -24,6 +26,143 @@ export type ObjectDropdownFilter = Omit<Filter, 'definition'> & {
+  let emptyRecordFilter: RecordGqlOperationFilter = {};
+
+  switch (filterType) {
+    case 'TEXT':
+    case 'EMAIL':
+    case 'PHONE':
```
```suggestion
    case filterType.TEXT:
    case filterType.EMAIL:
    case filterType.PHONE:
    ...
```

martmull replied:
forget that comment, FilterType is not an enum

martmull commented on packages/twenty-front/src/modules/object-record/object-filter-dropdown/utils/getOperandsForFilterType.ts:
```diff
@@ -5,25 +5,40 @@ import { FilterType } from '../types/FilterType';
+        ViewFilterOperand.DoesNotContain,
+        ...emptyOperands,
+      ];
     case 'CURRENCY':
```
`CURRENCY` isEmpty filter returns an error 
![image](https://github.com/twentyhq/twenty/assets/29927851/8a791bb0-38e0-4cff-8dbd-d67adb8a6fcb)

martmull replied:
NUMBER / INTEGER isEmpty returns graphql error
![image](https://github.com/twentyhq/twenty/assets/29927851/b9d6c9e7-d0eb-4b3f-bec2-51b53737d162)

martmull replied:
RELATION isEmpty returns coffee break error
![image](https://github.com/twentyhq/twenty/assets/29927851/7a5d4d9b-260e-4656-a3c7-abcc71ea655e)

martmull replied:
@pacyL2K19

pacyL2K19 replied:
Checking this asap @martmull 
Thanks for catching this

pacyL2K19 commented:
@martmull 
All the feedback implemented, please have a look
For the number filter, had to change the adapter on the backend, which resolved it, kindly let me know if anything else

martmull commented:
Hey @pacyL2K19, I just pulled and tested your branch. 

I have created an object with all the field types available:
![Pasted Graphic 7](https://github.com/twentyhq/twenty/assets/29927851/c961a51e-093d-4eea-ba77-e18400f4e8c2)

Then I have created 2 records, one empty, the other not empty, here are some issues I found:

## Missing filters
Some filters are missing in the filter dropdown. It is not related directly to your PR, let me know if you want me to fix this
- `boolean` 
- `Date`
- `Json`
- `Phone`
- `Rating`
- `Uuid`
![Pasted Graphic](https://github.com/twentyhq/twenty/assets/29927851/0349976f-a30b-43e6-883c-ea593cfa7231)

## `Address` Is Empty / Is not Empty not working
![Pasted Graphic 1](https://github.com/twentyhq/twenty/assets/29927851/212312fe-6903-4ab7-8c70-b3e3beecfdb5)
![Pasted Graphic 2](https://github.com/twentyhq/twenty/assets/29927851/6350f26f-2400-49d9-8d96-08c24f60c9e4)

## `Links` Is Empty / Is not Empty not working
![image](https://github.com/twentyhq/twenty/assets/29927851/cf000d5d-f43e-40ec-b84c-ab70d10806e0)
![Pasted Graphic 3](https://github.com/twentyhq/twenty/assets/29927851/a95500b5-db1e-4d02-96f3-b8ca6b20879e)

## `Multi-select` is empty / is not empty filter breaks server
![Pasted Graphic 5](https://github.com/twentyhq/twenty/assets/29927851/f0c2f9d4-aea8-44e3-820a-bf95baf80012)

## Missing isEmpty filter for `select` type 
![Pasted Graphic 6](https://github.com/twentyhq/twenty/assets/29927851/3907a329-a2ea-44b2-8f1f-4eb6220bf38d)

I know this PR lags behind a little bit, hope you are not tired of those PR comments! You already did a great job. Please let me know if you want me to fix those changes

martmull reviewed (changes requested):
see pr comment for functionnal issues

pacyL2K19 commented:
> see pr comment for functionnal issues

Thanks for testing out @martmull 
I haven't supported all the fields, the ticket description wasn't specific about all
I will work on it later today, we're getting there 😄 
I will ping you after pushing changes

pacyL2K19 commented:
@martmull 

I pushed some fixes:
- Added support of SELECT fields
- Fixed the Address filter logic: previously we were only checking for `addressStreet1`, `addressStreet2` and `addressCity`,  (that's why for France it wasn't returning data) so I added  the ability to check for `addressState`, `addressCountry` and `addressPostcode`, kindly let me know if this is a valid business rule
- Fixed the logic on Links fields, should be fine now

For the rest of fields you mentioned:

- boolean
- Date
- Json
- Phone
- Rating
- Uuid
- [AND] Multiselect
We currently don't support them anywhere, I suggest that we create a separate issue for this work, I volunteer to work on it (cc: @FelixMalfait) plus some details like:
- For the Multiselect fields we can support: Contains filters and empty filters,
- Or for the boolean IsFilters, and empty filter, etc 

@martmull @FelixMalfait  please let me know if this works for you so this PR doesn't go out of control 😃

martmull commented on packages/twenty-front/src/modules/object-record/object-filter-dropdown/utils/getOperandsForFilterType.ts:
```diff
@@ -5,24 +5,39 @@ import { FilterType } from '../types/FilterType';
+      return [...relationOperands, ...emptyOperands];
     case 'SELECT':
-      return [ViewFilterOperand.Is, ViewFilterOperand.IsNot];
+      return [...relationOperands];
```
i finally removed that as select cannot be empty

martmull reviewed (approved):
THank you!

pacyL2K19 commented:
> THank you!

Thanks for the review @martmull , if any additional issue to support other fields, please assign it to me
